server/services/userAdminService.js

```diff
@@ -111,49 +111,92 @@ module.exports = function createUserService(nomisClientBuilder, userClient) {
+
+    const bookingIds = await userClient.getCasesRequiringRo()
+    if (isEmpty(bookingIds)) {
+      return []
+    }
+
+    const all = await findRequiredRos(systemTokens.token, bookingIds)
+    if (isEmpty(all)) {
       return []
     }
 
-    const bookingIds = incomplete.map(u => u.bookingId)
-    const offenders = await getOffenderNomis(token, bookingIds)
+    const required = uniqueByAssignedId(all)
+    const incomplete = await findIncomplete(required)
+    if (isEmpty(incomplete)) {
+      return []
+    }
 
-    return addMappingDetails(incomplete, offenders)
+    return addOffenderNomis(systemTokens.token, incomplete)
   }
 
-  async function getOffenderNomis(token, bookingIds) {
-    try {
-      const nomisClient = nomisClientBuilder(token)
-      const offenders = await nomisClient.getOffenderSentencesByBookingId(bookingIds, false)
+  async function findRequiredRos(token, bookingIds) {
+    const required = await Promise.all(
+      bookingIds.map(async bookingId => {
+        try {
+          const { com } = await prisonerService.getCom(bookingId, token)
+          if (com) {
+            return { assignedId: com.deliusId, assignedName: com.name, bookingId }
+          }
+        } catch (error) {
+          logger.warn(`Error in getCom for incomplete users, booking id ${bookingId}`, error.stack)
+        }
+      })
+    )
 
-      if (isEmpty(offenders)) {
-        return null
-      }
+    return required.filter(e => e)
+  }
 
-      return new Map(offenders.map(o => [o.bookingId.toString(), o.offenderNo]))
-    } catch (error) {
-      logger.warn('Error getting offender nomis numbers for incomplete ROs', error.stack)
-      return []
-    }
+  function uniqueByAssignedId(assignments) {
+    return Object.values(
+      assignments.reduce((unique, assignment) => {
+        if (assignment.assignedId && !unique[assignment.assignedId]) {
+          return merge(unique, { [assignment.assignedId]: assignment })
+        }
+        return unique
+      }, {})
+    )
   }
 
-  function addMappingDetails(incomplete, offenderIds) {
-    return incomplete.map(u => ({
-      ...u,
-      offenderNomis: offenderIds ? offenderIds.get(u.bookingId.toString()) : undefined,
-      mapping: getMappingDetails(u),
-    }))
+  async function findIncomplete(assignments) {
+    const incomplete = await Promise.all(
+      assignments.map(async assignment => {
+        try {
+          const staff = await userClient.getRoUserByDeliusId(assignment.assignedId)
+          if (staff && !staff.onboarded) {
+            return merge(merge(assignment, staff), { mapped: true })
+          }
+          if (!staff) {
+            return merge(assignment, { mapped: false })
+          }
+        } catch (error) {
+          logger.warn(
+            `Error in getRoUserByDeliusId for incomplete users, assigned Id ${assignment.assignedId}`,
+            error.stack
+          )
+        }
+      })
+    )
+
+    return incomplete.filter(e => e)
```

Comment:
Not sure what you are filtering here

Reply:
removing any undefined from the result of map. For each item in the list we get an object if the user is incomplete, otherwise nothing, so the output of the map could have undefineds mixed in

Reply:
```
return incomplete.filter(Boolean)
```

Reply:
OK, cool, I didn't know you could do that